README.md
![Logo][Logo]

# nodezoo-search

- __Lead:__ [Dean McDonnell][Lead]
- __Sponsor:__ [nearForm][]

A micro-service that provides search for [NodeZoo][]. This micro-service requires an
instance of elastic-search running in order to correctly function. If elastic-search
cannot be reached the service will self terminate.

If you're using this microservice, and need help, you can:

- Post a [github issue][],
- Tweet to [@nodezoo][],
- Ask on the [Gitter][gitter-url].

## Running
This micro-service can be ran as part of a complete system or as a single isolated
unit.

### As a complete system
A special system repository is available that runs the complete system using Docker
and Fuge.

- [Nodezoo: The complete system][System]


### Isolated mode
To make testing easier this micro-service can be ran in 'isolated' mode. This mode
allows testing over http using a well defined port. Please note isolated mode means
patterns are not exposed via mesh.

To run in isolated mode,

 - Clone this repository locally,
 - Run `npm install`,
 - Run `SEARCH_ELASTIC_HOST=YOUR_HOST npm start isolated`,

__Note:__ You will need to know the cost of your vm if using Docker.

A simple http service is supported and can be called using Curl or other Rest client.
The default port is `8060`. It can be changed using the `SEARCH_PORT` environment
variable.

```
curl -d '{"role":"search","cmd":"search","query":"hapi"}' http://localhost:8060/act
```

## Configuration

### Running Elastic
A running instance of elastic search is required to use this service. Assuming you
have docker installed.

  - Start your docker machine if required
  - Run `eval $(docker-machine env default)` to enable docker in your shell.
  - Run `docker-compose -f test/elastic.yml up` to start elastic

Please note, in isolation mode you need to pass the host ip of your docker-machine
if you are on an OS other than linux, obtain with `docker-machine ip default`.

### Environment Variables
Various settings can be changed using environment variables, see the list below for
all available variable names.


#### SEARCH_HOST
  - The host to listen on in isolated mode.
  - Defaults to `localhost`

#### SEARCH_PORT
  - The port to listen on in isolated mode.
  - Defaults to `8060` .

#### SEARCH_ELASTIC_HOST
  - The host elastic will listen on.
  - Defaults to `localhost`

#### SEARCH_ELASTIC_PORT
  - The port elastic will listen on.
  - Defaults to `9200` .

#### SEARCH_ISOLATED
  - Starts isolated mode.
  - Defaults to `false`.

## Messages Handled
This micro-service emits the following messages.

### `role:search,cmd:upsert`
Adds or inserts a record into elastic search

```js
seneca.act(`role:search,cmd:upsert`, {data: {name:'seneca', ...}})
```

### `role:search,cmd:search`
Applies the provided query to elastic-search and returns the results.

```js
seneca.act(`role:search,cmd:search`, {query: 'seneca'}, (err, reply) => {})
```

### `role:info,info:updated`
Aliases upsert, this pattern is an integration point for the nodezoo system. It
allows the info micro-service to upsert without specifically calling nodezoo-search.

```js
seneca.act(`role:info,info:updated`, {data: {name:'seneca', ...}})
```

## Messages Emitted
This micro-service emits no messages.


## Contributing
The [NodeZoo org][] encourages __open__ and __safe__ participation.

- __[Code of Conduct]__

If you feel you can help in any way, be it with documentation, examples, extra testing, or new
features please get in touch.

## License
Copyright (c) 2014 - 2016, Richard Rodger and other contributors.
Licensed under [MIT][].


[MIT]: ./LICENSE
[Code of Conduct]: https://github.com/nodezoo/nodezoo-org/blob/master/CoC.md
[nearForm]: http://www.nearform.com/
[NodeZoo org]: http://www.nodezoo.com/
[main repo]: https://github.com/rjrodger/nodezoo
[Lead]: https://github.com/rjrodger
[Logo]: https://raw.githubusercontent.com/nodezoo/nodezoo-org/master/assets/logo-nodezoo.png
[NodeZoo]: https://github.com/nodezoo/nodezoo
[System]: https://github.com/nodezoo/nodezoo-system
[JSonic]: https://github.com/rjrodger/jsonic
[github issue]: https://github.com/nodezoo/nodezoo-search/issues
[@nodezoo]: http://twitter.com/nodezoo
[gitter-url]: https://gitter.im/nodezoo/nodezoo-org


CHANGELOG.md
## 7 April 2016 - 6.1.0

* Added CHANGELOG.md.
* Results set refactored.
* GitHub user added to results set.
* Added isolated mode.
* Corrected issue with missing author.
* Updated README to correct errata.
* Removed toolbag

## 6 April 2016 - 6.0.0

* First release of the 6th major revision of nodezoo.
* Switched to using npm released packages instead of git checkouts.


.toolbagrc.js
'use strict'

const StatsCollector = require('toolbag-plugin-stats-collector')
const UdpReporter = require('toolbag-plugin-udp-reporter')

module.exports = function (defaults, callback) {
  callback(null, {
    plugins: [
      {
        plugin: UdpReporter,
        options: {
          id: 'udp reporter',
          socketType: 'udp4',
          port: 5001,
          host: 'localhost'
        }
      },
      {
        plugin: StatsCollector,
        options: {
          enabled: true,
          period: 1000,
          eventLoopLimit: 30,
          features: {
            process: true,
            system: true,
            cpu: true,
            memory: true,
            gc: false,
            handles: false,
            requests: false,
            eventLoop: true,
            meta: {
              tags: ['nodezoo-search']
            }
          }
        }
      }
    ]
  })
}


lib/search.js
'use strict'

var ElasticSearch = require('elasticsearch')

var opts = {
  elastic: {
    host: 'localhost',
    port: 9200
  },
  mapping: {
    index: 'modules',
    type: 'document',
    body: {
      properties: {
        name: {
          type: 'string'
        },
        detail: {
          type: 'object'
        }
      }
    }
  }
}

module.exports = function (options) {
  var seneca = this
  var extend = seneca.util.deepextend

  opts = extend(opts, options)
  opts.elastic.client = new ElasticSearch.Client({
    host: {
      host: opts.elastic.host,
      port: opts.elastic.port
    }
  })

  ensureElasticSearch((err) => {
    if (err) {
      seneca.log.error(err)
      process.exit()
    }
  })

  seneca.add('role:search,cmd:search', search)
  seneca.add('role:search,cmd:upsert', upsert)
  seneca.add('role:info,info:updated', aliasUpsert)

  return {
    name: 'nodezoo-search'
  }
}

function ensureElasticSearch (done) {
  var client = opts.elastic.client
  var index = client.indices

  index.exists({index: 'modules'}, (err, exists) => {
    if (err) return done(err)
    if (exists) return done()

    index.create({index: 'modules'}, (err) => {
      if (err) return done(err)

      index.putMapping(opts.mapping, (err) => {
        done(err)
      })
    })
  })
}

function search (msg, done) {
  var client = opts.elastic.client
  var payload = {
    index: 'modules',
    body: {
      query: {
        match: {
          name: msg.query
        }
      }
    }
  }

  client.search(payload, (err, reply) => {
    if (err) return done(err)

    var hits = reply.hits && reply.hits.hits
    var items = []

    if (hits) {
      for (var i = 0; i < hits.length; i++) {
        items.push(hits[i]._source.detail)
      }
    }

    done(null, {
      ok: true,
      items: items
    })
  })
}

function upsert (msg, done) {
  var data = msg.data
  var github = data.github || {}
  var travis = data.travis || {}
  var npm = data.npm || {}
  var name = data.name

  var detail = {
    name: name,
    travis: {
      connected: data.travis ? true : false,
      url: travis.url || '',
      buildState: travis.buildState || ''
    },
    github: {
      connected: data.github ? true : false,
      stars: github.stars || 0,
      watches: github.watches || 0,
      forks: github.forks || 0,
      url: github.url || '',
      user: github.user || ''
    },
    npm: {
      connected: data.npm ? true : false,
      url: npm.url || '',
      version: npm.latestVersion || '',
      author: npm.author ? npm.author.name || '' : '',
      description: npm.description || '',
      homepage: npm.homepage || ''
    }
  }

  var client = opts.elastic.client
  var index = client.indices
  var payload = {
    index: 'modules',
    type: 'document',
    id: name,
    body: {
      name: name,
      detail: detail
    }
  }

  index.exists({index: 'modules'}, (err, exists) => {
    if (err) return done(err)
    if (exists) return

    index.create({index: 'modules'}, (err) => {
      if (err) return done(err)

      index.putMapping(opts.mapping, (err) => {
        if (err) return done(err)
      })
    })
  })

  client.index(payload, (err, reply) => {
    if (err) return done(err)

    done(null, {
      ok: true,
      index: reply._index,
      version: reply._version,
      intial: reply.created
    })
  })
}

function aliasUpsert (msg, done) {
  this.act('role:search,cmd:upsert', {data: msg.data}, done)
}


srv/start.js
'use strict'

var Seneca = require('seneca')
var Mesh = require('seneca-mesh')
var Search = require('../lib/search')

var envs = process.env
var opts = {
  seneca: {
    tag: envs.SEARCH_TAG || 'nodezoo-search'
  },
  search: {
    elastic: {
      host: envs.SEARCH_ELASTIC_HOST || 'localhost',
      port: envs.SEARCH_ELASTIC_PORT || '9200'
    }
  },
  mesh: {
    auto: true,
    listen: [
      {pin: 'role:search,cmd:upsert', model: 'consume'},
      {pin: 'role:search,cmd:search', model: 'consume'},
      {pin: 'role:info,info:updated', model: 'observe'}
    ]
  },
  isolated: {
    host: envs.SEARCH_HOST || 'localhost',
    port: envs.SEARCH_PORT || '8060'
  }
}

var Service =
Seneca(opts.seneca)
  .use(Search, opts.search)

envs.SEARCH_ISOLATED
  ? Service.listen(opts.isolated)
  : Service.use(Mesh, opts.mesh)
